rosidl_cli/rosidl_cli/command/generate/helpers.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2021 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import contextlib
+import json
+import os
+import pathlib
+import tempfile
+
+
+def package_name_from_include_path(path):
+    """
+    Derive ROS package name from a ROS interface dependency include path.
+
+    This function assumes ROS interface definition files follow the typical
+    ``rosidl`` install space layout i.e. 'package_name/subfolder/interface.idl'.
+    """
+    return pathlib.Path(path).resolve().parents[1].name
```

Comment:
It's a little surprising to me that the include path for idl files is slightly different from include paths for a C++ compiler.

With C++ the path to a header might be `/usr/include/package_name/foobar.hpp`, the include path given to the compiler would be `/usr/include` and the include in code would be `#include <package_name/foobar.hpp>`. Here the idl path of a dependency might be `/opt/ros/rolling/share/package_name/msg/foobar.idl`, the include path given to `rosidl generate --include-paths` would be `/opt/ros/rolling/share/package_name`, and the include using this idl in another idl (idl include syntax happens to look like C/C++) would be `#include "package_name/msg/foobar.idl"`.   With the `idl` path the last directory (`package_name/`) is both in `--include-path` and in `#include`.

I'm tempted to ask if the path passed can be made more like C++ include paths (`rosidl generate --include-paths` is `/opt/ros/rolling/share`), but then how would the generator know which packages in that folder are actually dependencies? Would changing that mean there would have to be another source of package dependencies other than the include paths? If so, maybe documenting that the last directory must be the package name in the `--include-paths` help text is an easier way forward.

Reply:
Hmm, there are no constraints on include paths. I think the confusion stems from `package_name _from_include_path()` awful naming. It expects a full file path, so it should've been called `package_name_from_interface_file_path()` :facepalm:.

Now, as you've pointed out, that lack of constrains means that for merged workspace installs there would be no way to tell which packages are dependencies and which are not. That's not a problem for any of the existing implementations (see only include paths' use case [here](https://github.com/ros2/rosidl_typesupport_connext/blob/7ae141c997f05c8d691e594680e89418be874b87/rosidl_typesupport_connext_cpp/rosidl_typesupport_connext_cpp/__init__.py#L26-L35), which literally reverses the conversion we do here from include paths to dependencies when writing generator arguments), except for the fact that generator arguments files will grow in size as it gets populated with IDL files. That's certainly not ideal, but not a fundamental flaw in the tool. Just a side effect of making thin wrappers instead of refactoring generators (which we can do later on).

Reply:
> I think the confusion stems from package_name_from_include_path()

Ah oops, yup. I read the name and didn't read how the path passed to it was from the output of `glob('**/*.idl')`. My mistake.